src/cmd_repl.rs

```diff
@@ -0,0 +1,142 @@
+    }
+
+    fn render_prompt_right(&self) -> Cow<str> {
+        Cow::default()
+    }
+
+    fn render_prompt_indicator(
+        &self,
+        _prompt_mode: reedline::PromptEditMode,
+    ) -> Cow<str> {
+        // not getting fancy for now
+        Cow::from("> ")
+    }
+
+    fn render_prompt_multiline_indicator(&self) -> Cow<str> {
+        Cow::from("... ")
+    }
+
+    fn render_prompt_history_search_indicator(
+        &self,
+        history_search: reedline::PromptHistorySearch,
+    ) -> Cow<str> {
+        let prefix = match history_search.status {
+            PromptHistorySearchStatus::Passing => "",
+            PromptHistorySearchStatus::Failing => "failing ",
+        };
+
+        Cow::Owned(format!(
+            "({}reverse-search: {}) ",
+            prefix, history_search.term
+        ))
+    }
+}
+
+fn repl(context: &mut humility::ExecutionContext) -> Result<()> {
+    // hardcoding to 100 entries for now. Even though this is called
+    // FileBackedHistory, this constructor only uses one in-memory.
+    let history = Box::new(FileBackedHistory::new(100));
+
+    let mut line_editor = Reedline::create()?.with_history(history)?;
+
+    let prompt = Prompt;
+
+    println!("Welcome to the humility REPL! Try out some subcommands, or 'quit' to quit!");
+
+    loop {
+        let sig = line_editor.read_line(&prompt)?;
+
+        match sig {
+            Signal::Success(buffer) => {
+                let buffer = buffer.trim();
+                match buffer {
+                    "quit" => {
+                        println!("Quitting!");
+                        std::process::exit(0);
+                    }
+                    "history" => {
+                        line_editor.print_history()?;
+                    }
+                    user_input => {
+                        let result = eval(context, user_input)?;
+                        println!("{result}");
+                    }
+                }
+            }
+            Signal::CtrlD | Signal::CtrlC => {
+                println!("\nAborted!");
+                break Ok(());
+            }
+            Signal::CtrlL => {
+                line_editor.clear_screen().unwrap();
+            }
+        }
+    }
+}
+
+fn eval(
+    context: &mut humility::ExecutionContext,
+    user_input: &str,
+) -> Result<String> {
+    let mut input = vec!["humility"];
+    input.extend(user_input.split(' '));
+
+    let (commands, _, cli) = match crate::parse_args(input) {
+        Some(s) => s,
+        None => return Ok(String::new()),
+    };
+
+    if let Some(s) = version(&cli)? {
+        return Ok(s);
+    }
+
+    // merge the new subcommand into our existing CLI, so that we retain options
+    // like the archive/dump, so that the user doesn't have to type them in again
+    context.cli.cmd = cli.cmd;
+
+    if let Err(e) = cmd::subcommand(context, &commands) {
+        Ok(format!(
+            "I'm sorry, Dave. I'm afraid I can't understand that: '{e}'",
```

Comment:
This is cute,  but ultimately distracting. For example:

```console
humility> test
I'm sorry, Dave. I'm afraid I can't understand that: 'does not appear to be a test archive'
```
I think we probably should have something more pedestrian here that more closely approximates what one see when running the command without the REPL (e.g., `humility; {cmdname} failed: {errmsg}`)

Reply:
I've pushed a change that simply shows the error, without the cutesy message. Output looks like this:

```
humility> tasks -f
error: Found argument '-f' which wasn't expected, or isn't valid in this context

        If you tried to supply `-f` as a value rather than a flag, use `-- -f`

USAGE:
    humility tasks [OPTIONS] [TASK]

For more information try --help
```

Given that it already mentions the command name in this output, I wonder if this is acceptable. If you want me to have it in the output anyway, I can do that, it'll just require moving some more code around (the command name isn't parsed out until further down the call stack).

Reply:
(And yeah I did think it was cutesy but not really useful for the final form, so might as well get rid of this here :) )

Reply:
Looks great -- thank you!